openssl s_server: don't use sendto() with connected UDP socket

ara4n opened the pull request:
Fixes https://github.com/openssl/openssl/issues/7675

On macOS, if you call connect() on a UDP socket you cannot then
call sendto() with a destination, otherwise it fails with Err#56
('socket is already connected').

By calling BIO_CTRL_DGRAM_SET_CONNECTED on the wbio we can tell
bio that the socket has been bound and make it call send() rather
than sendto().

CLA: Trivial

mattcaswell commented on apps/s_server.c:
```diff
@@ -2747,12 +2747,14 @@ static int init_ssl_connection(SSL *con)
-                    BIO_ADDR_free(client);
-                    return 0;
+                } else {
+                    if (BIO_connect(fd, client, 0) == 0) {
```
This change makes no sense. `fd` is -1 here, so the `BIO_connect` call is always going to fail? Also the logic before immediately errored out if `wbio == NULL`, but here you only attempt the connect if it is NULL (which shouldn't normally happen) - so we never normally hit this whole `else` branch.

ara4n replied:
i literally have no idea what i was thinking; sorry.

mattcaswell commented on apps/s_server.c:
```diff
@@ -2747,12 +2747,14 @@ static int init_ssl_connection(SSL *con)
+                        return 0;
+                    }
+                    wbio = SSL_get_wbio(con);
+                    BIO_ctrl(wbio, BIO_CTRL_DGRAM_SET_CONNECTED, 0, client);
```
This looks like it's going to fail. We know that `SSL_get_wbio` is going to return `NULL` (nothing has changed in `con` since it last returned `NULL` with this call), so the `BIO_ctrl` call will just immediately fail.

ara4n replied:
done.

mspncp commented:
Also, please fix the following issues of the subject line

* We don't usually place the `fix #xxxx` tag in the subject line. The `Fixes #7675` and the `CLA: trivial` tags need to go into the _body_ of the commit message on a separate line each.
* The formulation "marking connect()ed UDP sockets as connected": is a nice play with words but not very informative.

My suggestion:

```
openssl s_server: don't use sendto() with connected UDP socket

Fixes #7675

On macOS, if you call connect() on a UDP socket you cannot then
call sendto() with a destination, otherwise it fails with Err#56
('socket is already connected').

By calling BIO_CTRL_DGRAM_SET_CONNECTED on the wbio we can tell
bio that the socket has been bound and make it call send() rather
than sendto().

CLA: trivial
```

(Of course, if I am right w.r.t `BIO_ctrl_dgram_connect()`, then the commit message is due to change again.)


Apropos: I doubt whether this pull request qualifies as trivial.

ara4n commented:
@mattcaswell @mspncp thanks for the review; looks like my 2am brain totally thinkoed on the flow of control here; sorry for wasting time. Before I fix, can someone clarify whether using `BIO_ctrl_dgram_connect()` is the right solution or not?  (Also, is a ~oneliner like this considered trivial or not in the eyes of your CLA?)

mspncp commented:
Both are question which Matt can answer better than I do. But I'd like to weeken my "I doubt whether" to "I'm not sure whether" :-).

mspncp commented:
Matt, I am thinking of something like this here. What do you think about it (disclaimer: it compiles, but I didn't test it.) 

```diff
diff --git a/crypto/bio/bss_dgram.c b/crypto/bio/bss_dgram.c
index d5fe5bb5a8..63ebad7823 100644
--- a/crypto/bio/bss_dgram.c
+++ b/crypto/bio/bss_dgram.c
@@ -446,7 +446,18 @@ static long dgram_ctrl(BIO *b, int cmd, long num, void *ptr)
         ret = 1;
         break;
     case BIO_CTRL_DGRAM_CONNECT:
-        BIO_ADDR_make(&data->peer, BIO_ADDR_sockaddr((BIO_ADDR *)ptr));
+        if (!data->connected) {
+            BIO_ADDR_make(&data->peer, BIO_ADDR_sockaddr((BIO_ADDR *)ptr));
+            if (BIO_connect(b->num, &data->peer, 0) == 1) {
+                data->connected = 1;
+                ret = 0;
+            } else {
+                ret = -1;
+            }
+        } else {
+            /* TODO: BIOerr(...), because socket is already connected?*/
+            ret = -1;
+        }
         break;
         /* (Linux)kernel sets DF bit on outgoing IP packets */
     case BIO_CTRL_DGRAM_MTU_DISCOVER:
```

mattcaswell commented:
I don't think you can just co-opt `BIO_ctrl_dgram_connect()` for a different purpose. Although we don't currently use that internally, it is part of the public API so other people may be using it. We *could* invent a new CTRL for that purpose, although I'm not entirely sure it is worth it.

mattcaswell commented:
WRT, whether this is trivial or not we require that both reviewers and the original author all agree that a contribution is trivial from a *copyright* perspective. It doesn't take much for a change to be non-trivial but there are no hard and fast rules for determining it. If a contribution has creative content then it is non-trivial. So for example fixing a spelling mistake is typically trivial: there is only one way to do it. That is a test I often use for determining triviality: is there more than one way to do this? The fact that we are debating which way to do it suggests that whatever the final result is will be non-trivial.

mattcaswell commented:
> I disagree: It is neither used nor documented, so only the function signature is currently part of our public api, not the semantics.

This is just wrong. There are plenty of things that are in our public API that are not used internally. There are also plenty of things that in our public API that are not documented. Whether it is used and whether it is documented are not indicators for whether it is in the API. *Anything* that is in the public headers is considered part of the public API.

It is also not true that the semantics of functions does not form part of the API. It *does*. There are a whole stack of functions that, over the years, I have wished worked differently to the way they do - but we can't change them due to the potential breakage that might cause.

> And I seriously doubt that anybody digged into the source code to figure out the semantics and is currently using it. 

I am constantly surprised by the obscure stuff that some people are using from our API. We cannot make assumptions about this.

> But If you want to be sure, we could post a question to openssl-users.

An affirmative response to this would tell us there are definitely people using it. No response could still mean there are people using it: most of our users do not subscribe to openssl-users; even the ones that are may not recall without digging into their applications what obscure ctrls they have used.

mspncp commented:
My proposal for a new name would be `BIO_ctrl_connect()`, which matches the existing name `BIO_ctrl_set_connect()`.

Also, for consistency reasons I would propose to not only add a new constant `BIO_CTRL_CONNECT`, but also an alias `BIO_CTRL_SET_CONNECTED` with the same value as `BIO_CTRL_DGRAM_SET_CONNECTED` (and to deprecate the latter).

ara4n commented:
> The only thing that makes me hesitate is adding to the API unnecessarily. 

This was also my concern here. Is it really worth having a new API that calls connect() on a socket and then marks it as connected rather than just using the existing semantics that `BIO_connect()` and `BIO_ctrl_set_connected()` provide?  It feels like the API & doc overhead is worse than adding a note that says "N.B. if you call BIO_connect() on a datagram socket, you must also call BIO_ctrl_set_connected()".

mspncp commented:
Do as you prefer. It‘s a pity that the original implementation remained incomplete, but it is not your duty to come up with a better one.

ara4n commented:
i've removed the spectacular flow-of-control thinkos and added & tested the necessary `BIO_ctrl_set_connected`, and emailed over a CLA.

@mattcaswell ptal

mspncp commented:
I was too early with my approval, the CI bots complain about an unchecked return value:

```
apps/s_server.c:2758:17: error: value computed is not used [-Werror=unused-value]
                 BIO_ctrl_set_connected(wbio, client);
```

Note: since the return value is ignored in the other location, I won't complain if you just do it the same way:

https://github.com/openssl/openssl/blob/2a6f57bc098cdfa6320189f2725337c7f74a052e/apps/s_client.c#L2052

ara4n commented:
@mspncp rebased & force-pushed.

mspncp commented:
Cool! I've never seen this [force-pushed](https://github.com/openssl/openssl/compare/8caa1c4370b223242cced19cc9a60d6fdf00c781..279e39a4b2c64e416d11696267a2faf93609f09e) link before, which just appeared above your last comment. It seems to be a new feature in GitHub.

mspncp commented:
Indeed, it's a [new feature](https://blog.github.com/changelog/2018-11-15-force-push-timeline-event).

mattcaswell commented:
Close/reopen to kick CLA bot.

mspncp commented:
Pushed to master (b92678f4e94e) and 1.1.1 (49fabf6b4d3b). Thank you very much!
